### ISSUE TITLE
unsafe {} expression

### ISSUE THREAD

**Issue by ntrel** [vlang/v#5655]
I've been wrapping unsafe pointer operations in an `unsafe` block (which is now required in master but WIP). It's best practice to make that block as small as possible, because we want the compiler to verify any code that the programmer thinks is memory-safe. Also the reader should be able to easily understand why an expression is unsafe, so ideally there would be no safe expressions inside an unsafe block.

I've had to make changes that either add new mutable variables or change existing immutable variables to mutable (best practice is to limit the number of mutable variables needed). This could be avoided if `unsafe {}` was a valid expression.

builtin/array.v pointer arithmetic:
```diff
+	mut data := byteptr(0)
+	unsafe {
+		data = byteptr(a.data) + start * a.element_size
+	}
 	l := end - start
 	res := array{
 		element_size: a.element_size
-		data: byteptr(a.data) + start * a.element_size
+		data: data
 		len: l
 		cap: l
 	}
```
Preferred:
```v
		data: unsafe { byteptr(a.data) + start * a.element_size }
```
strconv/format.v pointer indexing:
```diff
-	if tmp.str[tmp.len-1] == `.` {
+	mut dot := bool(0)
+	unsafe 
+	{
+		dot = tmp.str[tmp.len-1] == `.`
+	}
+	if dot {
 		return tmp[..tmp.len-1]
 	}
```
Preferred:
```v
	if unsafe { tmp.str[tmp.len-1] } == `.` {
```
There are lots of these kind of changes, starting in #5581 and referenced issues.

Any tips on implementing `UnsafeExpr` would be appreciated.

**Comment by danieldaeschle** [comment]
@medvednikov do you want this?

**Comment by medvednikov** [comment]
Yeah, I think it's good.

**Comment by medvednikov** [comment]
@ntrel look at ExprStmt implementation, it's very simple and similar.

**Comment by spytheman** [comment]
I think if you want to wrap a single expression, then it should be:
`val := unsafe( expr )`
The `unsafe{ statements }` form can be used for blocks, just like it already is.

**Comment by ntrel** [comment]
Consider:
```v
p := unsafe {tmp := C.fun(); tmp + 1}
```
You can't write that if multiple statements aren't allowed, without violating best practices. This is the safest and most maintainable way to express that.

**Comment by spytheman** [comment]
If that is going to work, cool.

**Comment by ntrel** [comment]
Update:
1. I've realized that allowing statements inside an unsafe expression is quite difficult to implement correctly. E.g. `func(x, unsafe {x++ x++})` - really the only way I've thought of to do it right is to call a nested anonymous function `func(x, fn(){x++ return x++}())` - but I think nesting isn't supported yet. That would also mean that unsafe expressions were potentially inefficient if the optimizer doesn't inline the function call.
2. `{}` braces inside expressions looks a bit weird, e.g. `unsafe {&T(expr)}.len`.

So instead I have a branch that implements `unsafe(expr)`. However I think that `UnsafeStmt` could produce a result from it's last statement, to support multiple statements (as mentioned earlier in the discussion):
```v
p := unsafe {
  stmt
  expr
}
```

**Comment by joe-conigliaro** [comment]
https://github.com/vlang/v/pull/5973#issuecomment-664890285

**Comment by ntrel** [comment]
So Alex has decided to revert #5973:
https://github.com/vlang/v/pull/6015#issuecomment-665866413

I'm working on parsing `unsafe {stmt}` as an expression when `stmt` is an `ExprStmt`, which addresses the problems that were solved by #5973. It needs refactoring of `Parser.expr()` though.

**Comment by ntrel** [comment]
> which addresses the problems that were solved by #5973

The existing problems in master are all when `unsafe {expr}` is not treated as an expression. My fix is essentially done, I just need to clean up the early commits tomorrow (which were from the `unsafe(expr)` code):
https://github.com/vlang/v/compare/master...ntrel:unsafe-parse
